ce/api/streamflow/watershed.py

```diff
@@ -0,0 +1,421 @@
+    (Specifically, data *files* are not required, only the relevant contents
+    of those files passed as `DataGrid` objects. `DataGrid`s are much easier to
+    construct for tests.)
+
+    :param station_lonlat: (tuple) Location of drainage point, (lon, lat)
+    :param flow_direction: (DataGrid) Flow direction grid
+    :param elevation: (DataGrid) Elevation grid
+    :param area: (DataGrid) Area grid
+    :return: dict representation for JSON response object with the following
+        attributes:
+            area: Area of the watershed
+            elevation: Minimum and maximum elevations
+            shape: A GeoJSON object representing the outline of the watershed;
+                a concave hull of the cell rectangles.
+            hypsometric_curve: Elevation-area histogram of the watershed
+    """
+    if hypso_params is None:
+        # Default parameters cover total range of BC elevations from
+        # sea level to highest point, Mt. Fairweather, at 4,663 m.
+        hypso_params = {
+            'bin_start': 0,
+            'bin_width': 100,
+            'num_bins': 46,
+        }
+    if not flow_direction.is_compatible(elevation):
+        raise ValueError(
+            'Flow direction and elevation do not have compatible grids')
+    if not flow_direction.is_compatible(area):
+        raise ValueError(
+            'Flow direction and area do not have compatible grids')
+
+    # Compute lonlats of watershed whose mouth is at `station`
+    # TODO: Refactor to accept a DataGrid?
+    direction_matrix = VIC_direction_matrix(
+        flow_direction.lat_step, flow_direction.lon_step
+    )
+
+    watershed_xys, watershed_debug = build_watershed(
+        flow_direction.lonlat_to_xy(station_lonlat),
+        flow_direction.values,
+        direction_matrix,
+        debug=True
+    )
+
+    # `watershed_lonlats`, `elevations`, and `areas` must all be ordered
+    # collections (not sets) because it is required (at minimum) that the
+    # coordinates (lonlats) for `elevations[i]` and `areas[i]` be equal for
+    # all `i`.
+    watershed_lonlats = \
+        [flow_direction.xy_to_lonlat(xy) for xy in watershed_xys]
+
+    #  Compute elevations at each lonlat of watershed
+    ws_elevations = elevation.get_values_at_lonlats(watershed_lonlats)
+
+    #  Compute area of each cell in watershed
+    ws_areas = area.get_values_at_lonlats(watershed_lonlats)
+
+    # Compute the elevation/area curve
+    cumulative_areas = hypsometry(ws_elevations, ws_areas, **hypso_params)
+
+    # Compute outline of watershed as a GeoJSON feature
+    outline = outline_cell_rect(
+        watershed_lonlats, flow_direction.lat_step, flow_direction.lon_step)
+
+    return {
+        'elevation': {
+            'units': elevation.units,
+            'minimum': min(ws_elevations),
+            'maximum': max(ws_elevations),
+        },
+        'area': {
+            'units': area.units,
+            'value': sum(ws_areas),
```

Comment:
I had envisioned having `shapely` compute the area from the outline, but I guess this works too.